utils/waitable_counter.hh

```diff
@@ -0,0 +1,138 @@
+/*
```

Comment:
It looks like a very generic way to solve a very specific problem. This problem seems to be exactly like the problem a timer is solving (with a custom clock source).
This also means that there is no need to check the condition for every waiter since you have a comparator and you 
can preorder the wait queue.
It even aligns with the assumption of monotonicity of the counter.
I am not sure how important it is performance wise but maybe it is more correct conceptually. (?)
Maybe we even have another premade utility for this?

Reply:
It is more like a template for a very typical cond-usage, i.e. cond.wait([] { <some value > x> }).

Reply:
> t is more like a template for a very typical cond-usage, i.e. cond.wait([] { x> }).

I agree,  a timer mechanism can use what you described (it is a private case), however this typical usage can have optimizations
(at least for a long queue) - for example, instead of a list you can save the waiters in a priority heap and instead of
broadcast, only pop and signal until the waiter is below watermark. It also means that a wakeup is either an
exception condition (including timeout) or the cond being met.

What I meant in my comment is that the generic approach don't "enjoy" possible optimization due to the specific
nature of the problem - monotonicity and a pedefined condition which can evarage this monotonicity.

Reply:
Definition of operator< and the coroutine waiter optimizations should be optimal enough (we don't technically wake up until our condition callback is satisfied). Avi did not like the initial, more explicitly "optimal" version of this type. And I agree, a simple specialization of cond-var usage is easier on the debugger eye. And the usage here is not even super critical in any way.